### PR TITLE
Rename `expr::window_function::WindowFunction` to `WindowFunctionDefinition`, make structure consistent with ScalarFunction

### LINKED ISSUE #8347: Rename `expr::window_function::WindowFunction` to `WindowFunctionDefintion` for consistency
### Is your feature request related to a problem or challenge?

Part of  https://github.com/apache/arrow-datafusion/issues/8045 we are working
to consolidate the function implementation in DataFusion to make it easier to
manage DataFusion's built in list of functions as well as ensure user defined
and built in functions have the same feature sets.

I think it is also important to keep the Expr representation consistent for
Aggregate and Window functions to make DataFusion easier to work with, as well
as to permit eventually applying the same unification of built in and user
defined functions to aggregates and window functions.

### Describe the solution you'd like

I would like to rename the `WindowFunction` in to https://github.com/apache/arrow-datafusion/blob/4c914ea1e5d3dc61f3552adcffb8356c90d73bac/datafusion/expr/src/window_function.rs#L35-L44 to `WindowFunctionDefinition` so we didn't have so many things named `WindowFunction` and to mirror the structure of ScalarFunction

Edited: ~I would like to unify `Expr::WindowFunction` and `Expr::WindowUDF` following the pattern in https://github.com/apache/arrow-datafusion/pull/8258 @2010YOUY01:~

### Describe alternatives you've considered

_No response_

### Additional context

I think this is a pretty good first issue because we have an existing pattern in https://github.com/apache/arrow-datafusion/pull/8258

### PR DESCRIPTION
**Pull request by edmondop**
## Which issue does this PR close?

Closes #8347 

## Rationale for this change

The idea is to have the ScalarFunctions and WindowFunctions consistent in structure with each other

### CODE REVIEW

File: datafusion/expr/src/window_function.rs [hunk 1]
```diff
@@ -32,7 +32,7 @@ use strum_macros::EnumIter;
 
 /// WindowFunction
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub enum WindowFunction {
+pub enum WindowFunctionDefinition {
```
Comment (haohuaijin) [inline datafusion/expr/src/window_function.rs:35]: can we move `WindowFunctionDefinition` into `datafusion/expr/src/expr.rs`, make this consist of `ScalarFunctionDefinition` and `AggregateFunctionDefinition` .

Comment (edmondop) [reply datafusion/expr/src/window_function.rs]: If I understood correctly, this [2f4ce8368ae5003648919bb1201f08a2dbe23d5f](https://github.com/apache/arrow-datafusion/pull/8382/commits/2f4ce8368ae5003648919bb1201f08a2dbe23d5f) should be it. There are two commits after this, one to merge a change in expr.rs in the main branch and the following one to reformat

Comment (haohuaijin) [reply datafusion/expr/src/window_function.rs]: @edmondop sorry for make you misunderstood, I mean we only move the code related to `WindowFunctionDefinition`
https://github.com/apache/arrow-datafusion/blob/e19c669855baa8b78ff86755803944d2ddf65536/datafusion/expr/src/window_function.rs#L34-L44
https://github.com/apache/arrow-datafusion/blob/e19c669855baa8b78ff86755803944d2ddf65536/datafusion/expr/src/window_function.rs#L69-L78
https://github.com/apache/arrow-datafusion/blob/e19c669855baa8b78ff86755803944d2ddf65536/datafusion/expr/src/window_function.rs#L166-L178
https://github.com/apache/arrow-datafusion/blob/e19c669855baa8b78ff86755803944d2ddf65536/datafusion/expr/src/window_function.rs#L227-L237
into `datafusion/expr/src/expr.rs`.

File: datafusion/expr/src/window_function.rs [hunk 2]
```diff
@@ -1,470 +1,482 @@
...
-    // function suggestion. See https://github.com/apache/arrow-datafusion/issues/8082
-    fn test_display_and_from_str() {
-        for func_original in BuiltInWindowFunction::iter() {
-            let func_name = func_original.to_string();
-            let func_from_str = BuiltInWindowFunction::from_str(&func_name).unwrap();
-            assert_eq!(func_from_str, func_original);
-        }
-    }
-}
+// // Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline datafusion/expr/src/window_function.rs:1]: It seems strange that this file is entirely commented out (maybe it was supposed to be removed?)

File: datafusion/expr/src/expr.rs [hunk 3]
```diff
@@ -393,6 +398,258 @@ impl ScalarFunction {
...
+            WindowFunctionDefinition::WindowUDF(fun) => fun.fmt(f),
+        }
+    }
+}
+
+/// A [window function] built in to DataFusion
+///
+/// [window function]: https://en.wikipedia.org/wiki/Window_function_(SQL)
+#[derive(Debug, Clone, PartialEq, Eq, Hash, EnumIter)]
+pub enum BuiltInWindowFunction {
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:456]: Maybe you could put `BuiltInWindowFunction` in `datafusion/expr/src/built_in_window_function.rs` to mirror   `BuiltInFunction`?  
https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/built_in_function.rs

**Review by alamb — COMMENTED** [review]
Thanks @edmondop  -- this is looking like a nice step in the right direction

File: datafusion/expr/src/window_function.rs (hunk 2 above)
Comment (edmondop) [reply datafusion/expr/src/window_function.rs:1]: Yup, this was me being sloppy :)

File: datafusion/expr/src/expr.rs (hunk 3 above)
Comment (edmondop) [reply datafusion/expr/src/expr.rs:456]: @alamb @haohuaijin have I understood correctly?
- `BuiltinWindowFunction` goes into `datafusion/expr/src/built_in_window_function.rs`
- `WindowFunction` becomes `WindowFunctionDefinition` and goes into `datafusion/expr/src/expr.rs` with its trait implementations
- public functions in `window_function` stays there

Comment (haohuaijin) [reply datafusion/expr/src/expr.rs:456]: > * `BuiltinWindowFunction` goes into `datafusion/expr/src/built_in_window_function.rs`
> * `WindowFunction` becomes `WindowFunctionDefinition` and goes into `datafusion/expr/src/expr.rs` with its trait implementations

yes, you are correct.
> * public functions in `window_function` stays there

maybe those public functions also can move to `datafusion/expr/src/built_in_window_function.rs`. cc @alamb

**Comment by alamb** [comment]
This PR looks like it needs some work so marking it as a draft until it is ready for review again

File: datafusion/expr/src/built_in_window_function.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline datafusion/expr/src/built_in_window_function.rs:1]: ❤️

**Review by alamb — APPROVED** [review]
Thank you @edmondop -- I think this is great

I think it is excellent that the structure of the window functions now mirrors the structure of the scalar functions -- it will make navigating the datafusion codebase much easier

**Comment by alamb** [comment]
Thanks again @edmondop

**Pull request merged**